File: jdroid-gradle-commons-plugin/src/main/groovy/com/jdroid/gradle/commons/tasks/CreateGitHubReleaseTask.java
package com.jdroid.gradle.commons.tasks;

import com.jdroid.github.IRepositoryIdProvider;
import com.jdroid.github.Release;
import com.jdroid.github.client.GitHubClient;
import com.jdroid.github.service.ReleaseService;

import java.io.IOException;

public class CreateGitHubReleaseTask extends AbstractGitHubTask {

	public CreateGitHubReleaseTask() {
		setDescription("Create a GitHub Release");
	}

	@Override
	protected void onExecute() throws IOException {
		GitHubClient client = createGitHubClient();

		String tagName = "v" + getProject().getVersion();

		IRepositoryIdProvider repositoryIdProvider = getIRepositoryIdProvider();
		ReleaseService releaseService = new ReleaseService(client);
		
		Release release = releaseService.getReleaseByTagName(repositoryIdProvider, tagName);
		if (release == null) {
			String releaseNotes = propertyResolver.getStringProp("GIT_HUB_RELEASE_NOTES");
			createRelease(releaseService, repositoryIdProvider, tagName, releaseNotes);
			log("GitHub release created: " + tagName);
		} else {
			getLogger().warn("Skipping " + tagName + " release creation because it already exists.");
		}
	}

	private void createRelease(ReleaseService releaseService, IRepositoryIdProvider repositoryIdProvider, String name, String body) throws IOException {

		Release release = new Release();
		release.setBody(body);
		release.setDraft(false);
		release.setName(name);
		release.setTagName(name);
		release.setPrerelease(false);
		release.setTargetCommitish(getExtension().getGitBranch());

		releaseService.createRelease(repositoryIdProvider, release);
	}
}

File: jdroid-gradle-commons-plugin/src/main/groovy/com/jdroid/gradle/commons/tasks/CreateGitHubReleaseTask.kt
package com.jdroid.gradle.commons.tasks

import com.jdroid.github.IRepositoryIdProvider
import com.jdroid.github.Release
import com.jdroid.github.client.RequestException
import com.jdroid.github.service.ReleaseService
import java.io.IOException

open class CreateGitHubReleaseTask : AbstractGitHubTask() {
    init {
        description = "Create a GitHub Release"
    }

    @Throws(IOException::class)
    override fun onExecute() {
        val client = createGitHubClient()

        val tagName = "v" + project.version

        val repositoryIdProvider = iRepositoryIdProvider
        val releaseService = ReleaseService(client)

        try {
            releaseService.getReleaseByTagName(repositoryIdProvider, tagName)
            logger.warn("Skipping $tagName release creation because it already exists.")
        } catch (e: RequestException) {
            if (e.status == 404) {
                val releaseNotes = propertyResolver.getStringProp("GIT_HUB_RELEASE_NOTES")
                createRelease(releaseService, repositoryIdProvider, tagName, releaseNotes)
                log("GitHub release created: $tagName")
            } else {
                throw e
            }
        }
    }

    @Throws(IOException::class)
    private fun createRelease(releaseService: ReleaseService, repositoryIdProvider: IRepositoryIdProvider, name: String, body: String) {

        val release = Release()
        release.body = body
        release.isDraft = false
        release.name = name
        release.tagName = name
        release.isPrerelease = false
        release.targetCommitish = extension.gitBranch

        releaseService.createRelease(repositoryIdProvider, release)
    }
}
